nexus/src/db/datastore/mod.rs

```diff
@@ -0,0 +1,1061 @@
+            Err(error) => {
+                public_error_from_diesel_pool(error, ErrorHandler::Server)
+            }
+        }
+    }
+}
+
+/// Constructs a DataStore for use in test suites that has preloaded the
+/// built-in users, roles, and role assignments that are needed for basic
+/// operation
+#[cfg(test)]
+pub async fn datastore_test(
+    logctx: &dropshot::test_util::LogContext,
+    db: &omicron_test_utils::dev::db::CockroachInstance,
+) -> (OpContext, Arc<DataStore>) {
+    use crate::authn;
+
+    let cfg = db::Config { url: db.pg_config().clone() };
+    let pool = Arc::new(db::Pool::new(&cfg));
+    let datastore = Arc::new(DataStore::new(pool));
+
+    // Create an OpContext with the credentials of "db-init" just for the
+    // purpose of loading the built-in users, roles, and assignments.
+    let opctx = OpContext::for_background(
+        logctx.log.new(o!()),
+        Arc::new(authz::Authz::new(&logctx.log)),
+        authn::Context::internal_db_init(),
+        Arc::clone(&datastore),
+    );
+    datastore.load_builtin_users(&opctx).await.unwrap();
+    datastore.load_builtin_roles(&opctx).await.unwrap();
+    datastore.load_builtin_role_asgns(&opctx).await.unwrap();
+    datastore.load_builtin_silos(&opctx).await.unwrap();
+    datastore.load_silo_users(&opctx).await.unwrap();
+    datastore.load_silo_user_role_assignments(&opctx).await.unwrap();
+
+    // Create an OpContext with the credentials of "test-privileged" for general
+    // testing.
+    let opctx =
+        OpContext::for_tests(logctx.log.new(o!()), Arc::clone(&datastore));
+
+    (opctx, datastore)
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::authn;
+    use crate::authz;
+    use crate::db::explain::ExplainableAsync;
+    use crate::db::fixed_data::silo::SILO_ID;
+    use crate::db::identity::Asset;
+    use crate::db::identity::Resource;
+    use crate::db::lookup::LookupPath;
+    use crate::db::model::Dataset;
+    use crate::db::model::InstanceExternalIp;
+    use crate::db::model::Rack;
+    use crate::db::model::Region;
+    use crate::db::model::Service;
+    use crate::db::model::SiloUser;
+    use crate::db::model::Sled;
+    use crate::db::model::SshKey;
+    use crate::db::model::VpcSubnet;
+    use crate::db::model::Zpool;
+    use crate::db::model::{ConsoleSession, DatasetKind, Project, ServiceKind};
+    use crate::db::queries::vpc_subnet::FilterConflictingVpcSubnetRangesQuery;
+    use crate::external_api::params;
+    use chrono::{Duration, Utc};
+    use nexus_test_utils::db::test_setup_database;
+    use omicron_common::api::external::{
+        ByteCount, Error, IdentityMetadataCreateParams, LookupType, Name,
+    };
+    use omicron_test_utils::dev;
+    use std::collections::HashSet;
+    use std::net::Ipv6Addr;
+    use std::net::SocketAddrV6;
+    use std::net::{IpAddr, Ipv4Addr, SocketAddr};
+    use std::sync::Arc;
+    use uuid::Uuid;
+
+    #[tokio::test]
```

Comment:
It'll probably make sense to move some of these out too, but we can punt on that for now